### PR TITLE
Infinite Staking Part 2

### PR DESCRIPTION
**Pull request by Doy-lee**
As a refresher, the user flow is as follows,
* Prepare registration with the Service Node daemon to get the requisite command
* Input the command into wallet-cli and submit the transaction, i.e stake (to yourself), key images are pre-calculated and submitted in the TX extra.
* Node is registered on the network whose expiration date is infinite and key images of the staked balance are locked on the network
* When the user wishes to spend their stake, they must use the new command request_stake_unlock <service_node_key> in the wallet CLI to submit the "Key Image Unlock" transaction
* The key images will be unlocked from the curr_height + 15 days worth of blocks
* During the stake period, staked funds do not show up as unlocked balance and cannot be spent.

Part 2 Includes
* Removing the ability to autostake and associated code
* Migration step for DB version 3 -> 4 to generate a table of blacklisted outputs to avoid for mixins on TX construction
* Changing ```request_stake_unlock``` duration to be half the duration of the blacklist duration. I.e. if blacklist blocks key images for 30 days, requesting to unlock, unlocks the key image in 15 days.
* Key Image Limit in the protocol is currently a global maximum of 4 key images per node, where each contribution must be at least ```((1/remaining key images) * staking_requirement)```
* Rework ```prepare_registration``` for better UX (https://github.com/loki-project/loki/issues/372) and accommodate the new changes in contribution rules. Users can cancel/go backwards and redo registration information at any point in the dialog.
* Add ```print_locked_stakes``` to query the key images currently locked in the wallet
* Wallet prompt UI now displays if there are any key images currently locked up in the Service Node network

### CODE REVIEW

**Comment by Doy-lee** [comment]
This is ready for review @msgmaxim

File: src/wallet/wallet2.cpp [hunk 1]
```diff
@@ -7514,7 +7601,26 @@ void wallet2::get_outs(std::vector<std::vector<tools::wallet2::get_outs_entry>>
       if (n_rct == 0)
         return rct_offsets[block_offset] ? rct_offsets[block_offset] - 1 : 0;
       MDEBUG("Picking 1/" << n_rct << " in " << (last_block_offset - first_block_offset + 1) << " blocks centered around " << block_offset + rct_start_height);
-      return first_rct + crypto::rand<uint64_t>() % n_rct;
+      
+      uint64_t pick = first_rct + crypto::rand<uint64_t>() % n_rct;
```
Comment (msgmaxim) [inline src/wallet/wallet2.cpp]: It is worth pointing out that % makes the resulting random distribution non-uniform. Not sure if we care much in this case.

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -3138,6 +3137,51 @@ bool wallet2::get_rct_distribution(uint64_t &start_height, std::vector<uint64_t>
   return true;
 }
 //----------------------------------------------------------------------------------------------------
+bool wallet2::get_output_blacklist(std::vector<uint64_t> &blacklist)
+{
+  uint32_t rpc_version;
+  boost::optional<std::string> result = m_node_rpc_proxy.get_rpc_version(rpc_version);
+  // no error
+  if (!!result)
```
Comment (msgmaxim) [inline src/wallet/wallet2.cpp]: Is there any reason to prefer this over `if(result)`?

File: src/blockchain_db/lmdb/db_lmdb.h [hunk 3]
```diff
@@ -329,6 +332,8 @@ class BlockchainLMDB : public BlockchainDB
   std::map<uint64_t, std::tuple<uint64_t, uint64_t, uint64_t>> get_output_histogram(const std::vector<uint64_t> &amounts, bool unlocked, uint64_t recent_cutoff, uint64_t min_count) const;
 
   bool get_output_distribution(uint64_t amount, uint64_t from_height, uint64_t to_height, std::vector<uint64_t> &distribution, uint64_t &base) const;
+  virtual bool get_output_blacklist(std::vector<uint64_t>       &blacklist) const;
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/blockchain_db/lmdb/db_lmdb.h]: Added. The others can be done in another commit

File: src/daemon/rpc_command_executor.cpp [hunk 4]
```diff
@@ -2531,272 +2614,437 @@ bool t_rpc_command_executor::prepare_registration()
...
+        if (last_input_result == input_line_result::back)
+          continue;
+
+        if (last_input_result == input_line_result::cancel)
+        {
+          step = register_step_cancelled_by_user;
+          continue;
+        }
+
+        int additional_contributors = atoi(input.c_str());
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/daemon/rpc_command_executor.cpp]: I've opted for strtol which just returns LONG_MAX or LONG_MIN on out of range so we can skip exception handling. https://en.cppreference.com/w/c/string/byte/strtol

File: src/daemon/rpc_command_executor.cpp [hunk 5]
```diff
@@ -2531,272 +2614,437 @@ bool t_rpc_command_executor::prepare_registration()
...
+        if (!cryptonote::parse_amount(contribution, contribution_str))
+        {
+          std::cout << "Invalid amount." << std::endl;
+          continue;
+        }
+
+        uint64_t portions = service_nodes::get_portions_to_make_amount(staking_requirement, contribution);
+        if (portions < min_contribution_portions)
+        {
+          std::cout << "Invalid amount." << std::endl;
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/daemon/rpc_command_executor.cpp]: Fixed

File: src/daemon/rpc_command_executor.cpp [hunk 6]
```diff
@@ -2531,272 +2614,437 @@ bool t_rpc_command_executor::prepare_registration()
...
-  }
 
-  if(is_solo_stake)
-  {
-    contributions.push_back(STAKING_PORTIONS);
-    portions_remaining = 0;
-    total_reserved_contributions += get_actual_amount(staking_requirement, STAKING_PORTIONS);
-  }
-  else
+  enum register_step
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/daemon/rpc_command_executor.cpp]: Fixed

File: src/simplewallet/simplewallet.cpp [hunk 7]
```diff
@@ -5865,125 +5827,14 @@ bool simple_wallet::register_service_node_main(
   bool submitted_to_network = false;
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/simplewallet/simplewallet.cpp]: Line removed

File: src/wallet/wallet2.cpp (hunk 1 above)
Comment (Doy-lee) [reply src/wallet/wallet2.cpp]: Was previously like that, so should be ok.

File: src/wallet/wallet2.cpp (hunk 2 above)
Comment (Doy-lee) [reply src/wallet/wallet2.cpp]: Copy pasta. I just made it ```if (result)``` now and updated some of the error messages there to be relevant for output blacklist.

**Pull request merged**